File: packages/template-php/tests/ObjectCallsTest.php
<?php

declare(strict_types=1);

namespace Polyspec\Template\Tests;

use PHPUnit\Framework\TestCase;
use Polyspec\Template\AstProgram;
use Polyspec\Template\Engine;
use Polyspec\Template\Loader\ArrayLoader;

final class ObjectCallsTest extends TestCase
{
    public function testAssignedObjectRetainsPublicFieldsAndMethods(): void
    {
        $program = new AstProgram(new ArrayLoader(['page.tpl' => '{= order.total}|{= order.status_label("ready")}']));
        $order = new class {
            public float $total = 12.0;
            public function status_label(string $prefix): string { return $prefix . ':' . (int) $this->total; }
        };

        self::assertSame('12|ready:12', (new Engine($program))->render('page.tpl', ['order' => $order]));
    }

    public function testRegisteredClassFunctionIsCallable(): void
    {
        $program = new AstProgram(new ArrayLoader(['page.tpl' => '{= Order::status_label("ready")}']));
        $program->registerClass('Order', 'status_label', static fn (array $args): string => $args[0] . ':ok');

        self::assertSame('ready:ok', (new Engine($program))->render('page.tpl'));
    }
}

File: packages/template-go/render/object_calls_test.go
package render

import (
	"testing"

	"github.com/polyspec/template/ast"
	"github.com/polyspec/template/functions"
	"github.com/polyspec/template/value"
)

type testOrder struct{ Total float64 }

func (o *testOrder) StatusLabel(prefix string) string { return prefix + ":12" }

func TestAssignedObjectRetainsPublicFieldsAndMethods(t *testing.T) {
	runtime, err := NewRuntimeEnvironment(nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	root := value.NewOrderedMap()
	root.Set("order", &testOrder{Total: 12})
	context := NewContext(runtime, root, functions.Env{Timezone: "Z"}, "page.tpl")
	bindings := NewRuntimeBindings(context)
	if got := bindings.Member(root.MustGet("order"), "total"); got != float64(12) {
		t.Fatalf("member = %v", got)
	}
	frame := NewFrame("page.tpl", nil, root)
	got, err := bindings.MemberCall(root.MustGet("order"), "status_label", []value.Value{"ready"}, frame, ast.Span{0, 1})
	if err != nil {
		t.Fatal(err)
	}
	if got != "ready:12" {
		t.Fatalf("method = %v", got)
	}
}

func TestRegisteredClassFunctionIsCallable(t *testing.T) {
	runtime, err := NewRuntimeEnvironment(nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	if err := runtime.RegisterClass("Order", "status_label", func(args []value.Value, _ functions.Context) (value.Value, error) {
		return args[0].(string) + ":ok", nil
	}); err != nil {
		t.Fatal(err)
	}
	root := value.NewOrderedMap()
	context := NewContext(runtime, root, functions.Env{Timezone: "Z"}, "page.tpl")
	got, err := NewRuntimeBindings(context).ClassCall("Order", "status_label", []value.Value{"ready"}, NewFrame("page.tpl", nil, root), ast.Span{0, 1})
	if err != nil {
		t.Fatal(err)
	}
	if got != "ready:ok" {
		t.Fatalf("class method = %v", got)
	}
}
